worker/controlsocket/handlers.go

```diff
@@ -0,0 +1,247 @@
+// Copyright 2023 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package controlsocket
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"strings"
+
+	"github.com/gorilla/mux"
+	"github.com/juju/errors"
+	"github.com/juju/names/v4"
+
+	"github.com/juju/juju/core/permission"
+	"github.com/juju/juju/environs/bootstrap"
+	"github.com/juju/juju/state"
+)
+
+const (
+	// jujuMetricsUserPrefix defines the "namespace" in which this worker is
+	// allowed to create/remove users.
+	jujuMetricsUserPrefix = "juju-metrics-"
+
+	// userCreator is the listed "creator" of metrics users in state.
+	// This user CANNOT be a local user (it must have a domain), otherwise the
+	// model addUser code will complain about the user not existing.
+	userCreator = "controller@juju"
+)
+
+func (w *Worker) registerHandlers(r *mux.Router) {
+	// We're using a fully REST-ful API structure here.
+	r.HandleFunc("/metrics-users", w.handleAddMetricsUser)
+	r.HandleFunc("/metrics-users/{username}", w.handleRemoveMetricsUser)
+}
+
+func (w *Worker) handleAddMetricsUser(resp http.ResponseWriter, req *http.Request) {
+	if req.Method != http.MethodPost {
+		resp.Header().Add("Allow", http.MethodPost)
+		w.writeResponse(resp, http.StatusMethodNotAllowed, errorf("method must be %q, got %q", http.MethodPost, req.Method))
+		return
+	}
+
+	body := &struct {
+		Username string `json:"username"`
+		Password string `json:"password"`
+	}{}
+
+	defer req.Body.Close()
+	err := json.NewDecoder(req.Body).Decode(body)
```

Comment:
I'd rather we just create the type outside of the method body. It doesn't provide any advantage inside the method.

```suggestion
	var body authBody

	defer req.Body.Close()
	err := json.NewDecoder(req.Body).Decode(&body)
```